zebra-network/src/peer/handshake.rs

```diff
@@ -36,43 +36,149 @@ use super::{Client, Connection, ErrorSlot, HandshakeError};
+    relay: Option<bool>,
+}
+
+impl<S> Builder<S>
 where
     S: Service<Request, Response = Response, Error = BoxedStdError> + Clone + Send + 'static,
     S::Future: Send,
 {
-    /// Construct a new `PeerConnector`.
-    pub fn new(
-        config: Config,
-        internal_service: S,
-        timestamp_collector: mpsc::Sender<MetaAddr>,
-    ) -> Self {
-        // XXX this function has too many parameters, but it's not clear how to
-        // do a nice builder as all fields are mandatory. Could have Builder1,
-        // Builder2, ..., with Builder1::with_config() -> Builder2;
-        // Builder2::with_internal_service() -> ... or use Options in a single
-        // Builder type or use the derive_builder crate.
-        Handshake {
+    /// Provide a config.  Mandatory.
+    pub fn with_config(mut self, config: Config) -> Self {
+        self.config = Some(config);
+        self
+    }
+
+    /// Provide a service to handle inbound requests. Mandatory.
+    pub fn with_inbound_service(mut self, inbound_service: S) -> Self {
+        self.inbound_service = Some(inbound_service);
+        self
+    }
+
+    /// Provide a hook for timestamp collection. Optional.
+    ///
+    /// If this is unset, timestamps will not be collected.
+    pub fn with_timestamp_collector(mut self, timestamp_collector: mpsc::Sender<MetaAddr>) -> Self {
+        self.timestamp_collector = Some(timestamp_collector);
+        self
+    }
+
+    /// Provide this node's address, to send to peers.  Optional.
+    ///
+    /// If this is unset, the default of 0.0.0.0:8233 will be used.
+    pub fn with_addr(mut self, addr: SocketAddr) -> Self {
+        self.our_addr = Some(addr);
+        self
+    }
+
+    /// Provide the services this node advertises to other peers.  Optional.
+    ///
+    /// If this is unset, the node will advertise itself as a client.
+    pub fn with_advertised_services(mut self, services: PeerServices) -> Self {
+        self.our_services = Some(services);
+        self
+    }
+
+    /// Provide this node's user agent.  Optional.
+    ///
+    /// This must be a valid BIP14 string.  If it is unset, the user-agent will be empty.
+    pub fn with_user_agent(mut self, user_agent: String) -> Self {
+        self.user_agent = Some(user_agent);
+        self
+    }
+
+    /// Whether to request that peers relay transactions to our node.  Optional.
+    ///
+    /// If this is unset, the node will not request transactions.
+    pub fn want_transactions(mut self, relay: bool) -> Self {
+        self.relay = Some(relay);
+        self
+    }
+
+    /// Consume this builder and produce a [`Handshake`].
+    ///
+    /// Returns an error only if any mandatory field was unset.
+    pub fn finish(self) -> Result<Handshake<S>, &'static str> {
+        let config = self.config.ok_or("did not specify config")?;
+        let inbound_service = self
+            .inbound_service
+            .ok_or("did not specify inbound service")?;
```

Comment:
We could use Session types to prevent this at compile time for no reason other than for fun

Reply:
we already have session types at home

session types at home: unwrapping a `&'static str`